<a href="https://colab.research.google.com/github/angemoreno/Diplomado/blob/main/EDA_consumo.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Primero se cargan las librerías y se cargan los datos

In [ ]:
import pandas as pd
import numpy as np
from statistics import mean

filename = "ds-energy-consumers.csv"
df = pd.read_csv(filename, sep=",")

In [ ]:
df

NIC  Estrato  Nhab  Ingresos  Consumo  IndReca  QSAT
0      1234686      1.0     5      3.62   286.00      0.2   2.4
1      1234779      1.0     2      1.55   236.00      0.2   1.1
2      1235077      4.0     6     16.72   304.44      0.4   3.0
3      1235104      2.0     4      5.63   296.00      0.2   4.7
4      1235131      4.0     5     13.96   390.00      0.4   1.8
...        ...      ...   ...       ...      ...      ...   ...
21912  2999555      3.0     4     11.21   352.00      0.3   1.5
21913  2999667      4.0     2      5.61   276.00      0.4   2.7
21914  2999716      2.0     5      7.07   321.00      0.2   2.4
21915  2999730      4.0     6     16.73   304.44      0.4   1.5
21916  2999862      2.0     3      4.32   273.00      0.2   2.7

[21917 rows x 7 columns]

Paso los valores negativos a positivos absolutizando con la función abs()

In [ ]:
df = abs(df)
df.describe()

NIC       Estrato          Nhab      Ingresos       Consumo  \
count  2.191700e+04  21917.000000  21917.000000  21917.000000  21917.000000   
mean   2.117604e+06      2.513711      4.006981      7.063782    304.436103   
std    5.108147e+05      1.111775      1.414729      4.089644     42.673752   
min    1.234686e+06      1.000000      2.000000      1.480000    235.000000   
25%    1.678706e+06      2.000000      3.000000      3.640000    272.000000   
50%    2.120192e+06      3.000000      4.000000      6.300000    302.000000   
75%    2.558549e+06      3.000000      5.000000     10.430000    344.000000   
max    2.999862e+06      4.000000      6.000000     16.730000    390.000000   

            IndReca          QSAT  
count  21917.000000  21917.000000  
mean       0.295613      2.949172  
std        0.093297      1.155652  
min        0.200000      1.000000  
25%        0.200000      2.000000  
50%        0.300000      2.900000  
75%        0.400000      3.900000  
max        0.400000      4.900000

In [ ]:
valores_nulos = df.isnull().sum() #Ya que la matriz es muy extensa para identificar los valores nulos, realizo la sumatorio de los valores faltantes por atributo
print(valores_nulos) #Defino la sumatoria de los valores nulos por atriuto como una variable "valores_nulos"

NIC         0
Estrato     0
Nhab        0
Ingresos    0
Consumo     0
IndReca     0
QSAT        0
dtype: int64


In [ ]:
(df == 0).sum() #Busco valores que se encuentren en 0 en las celdas para cada atributo, para esto a través de una lista de la suma de los atributos que tienen valores en

NIC         0
Estrato     0
Nhab        0
Ingresos    0
Consumo     0
IndReca     0
QSAT        0
dtype: int64

Se observa que tengo las siguientes inconsistencias:


*   Los valores del atributo Estrato deben estar entre los rangos 1 - 4, pero su valor máximo es 5, es decir, este atributo tiene valores atipicos.
*   El atributo Estrato tiene 395 valores nulos.
*   Los valores del atributo Consumo deben estar entre los rangos 200 - 400 kWh, pero su valor máximo es 99999, entonces, este atributo también tiene valores atipicos.
*    El atributo Consumo tiene 51 valores nulos.
*   Los valores del atributo IndReca deben estar entre los rangos 0 - 1 y estos valores deben estar entre el intervalo entre 25% y 75% de los valores. *****
*   El atributo IndReca tiene 48 valores nulos.
*   Los valores del atributo QSAT deben estar entre los rangos 1 - 5, sin embargo, este atributo tiene valores atipicos al tener un valor mínimo de 0 y valor mínimo de 0.
*   El atributo QSAT tiene 4 valores en cero.




## **Para el atributo Consumo**

Dado que los valores en Consumo tiene valores muy grandes que se salen del rango pedido por la empresa (200 - 420 kWh), al calcular el promedio, me dará un valor que también estará fuera del rango solicitado. Por lo que para resolver esto, los valores que están por fuera del rango se reemplazarán por "NaN", Not a Number, que es un valor float que se usará para representar valores indefinidos o no representables evitando que se afecte el valor promedio y posteriormente reemplazo los valores NaN por el promedio de Consumo.

In [ ]:
NaN = float("nan") # Creo la variable "NaN"
df.loc[((df["Consumo"]<200) | (df["Consumo"]>420)), "Consumo"] = NaN #Hallo los valores en Consumo fuera del rango y lo igualo a la variable NaN

In [ ]:
promedio_consumo= round(df["Consumo"].mean(), 2) #Saco el promedio para el atributo Consumo y lo redondeo a dos cifras significativas.

In [ ]:
df.loc[df["Consumo"].isnull(), "Consumo"] = promedio_consumo #Reemplazo los valores nulos de estrato con el valor promedio del atributo Consumo.

##**Para el atributo Estrato**

# Valores nulos de Estrato

In [ ]:
promedio_estrato = round(df["Estrato"].mean(), 0) #Hayo el promedio para el atributo Estrato y lo redondeo a dos cifras significativas.

In [ ]:
df.loc[df["Estrato"].isnull(), "Estrato"] = promedio_estrato #Reemplazo los valores nulos de estrato con el valor promedio del atributo Estrato.

# Valores atipicos de Estrato

In [ ]:
df[((df["Estrato"]<1) | (df["Estrato"]>4))] #Se sacan todos los valores atipicos del Estrato: menores a 1 y mayores a 4

Empty DataFrame
Columns: [NIC, Estrato, Nhab, Ingresos, Consumo, IndReca, QSAT]
Index: []

In [ ]:
df.loc[((df["Estrato"]<1) | (df["Estrato"]>4)), "Estrato"] = promedio_estrato # Reemplazo el valor atipico del Estrato: 5, por el valor promedio del atributo.

##**Para el atributo QSAT**

#Valores atípicos de QSAT

In [ ]:
df.loc[(df["QSAT"] == 0)] #Identifico las filas donde QSAT sea cero

Empty DataFrame
Columns: [NIC, Estrato, Nhab, Ingresos, Consumo, IndReca, QSAT]
Index: []

In [ ]:
promedio_QSAT= round(df["QSAT"].mean(), 2) #Hayo el promedio para el atributo QSAT y lo redondeo a dos cifras significativas.

In [ ]:
df.loc[df["QSAT"] == 0, "QSAT"] = promedio_QSAT #Reemplazo los valores nulos de estrato con el valor promedio del atributo IndReca.

##**Para el atributo IndReca**

# Valores nulos de IndReca

In [ ]:
promedio_IndReca= round(df["IndReca"].mean(), 2) #Hayo el promedio para el atributo IndReca y lo redondeo a dos cifras significativas.

In [ ]:
df.loc[df["IndReca"].isnull(), "IndReca"] = promedio_IndReca #Reemplazo los valores nulos de estrato con el valor promedio del atributo IndReca.

# Valores atipicos de IndReca

Ya que los valores de IndReca deben estar entre el 25% y el 75%, los valores atípicos serán reemplazados por el promedio entre 25-50% para valores por debajo del 25% y para valores superiores al 75% serán reemplazados por el promedio entre 50-75%. Para esto, primero usando la función "percentile" de numpy saco el 25, 50 y 75% del atributo IndReca del dataframe y los asigno como las variables Q1, Q2 y Q3 respectivamente. Luego, hayo los valores menores o iguales a 25% y los reemplazo por el promedio entre 25-50%. También lo realizo para los valores mayores o iguales a 75% y los reemplazo por el promedio entre 50-75%.

In [ ]:
# Nombro las variables Q1, Q2 y Q3 para los percentiles de 25, 50 y 75% respectivamente de el IndReca
Q1 = np.percentile(df.IndReca, 25)
Q2 = np.percentile(df.IndReca, 50)
Q3 = np.percentile(df.IndReca, 75)

In [ ]:
df.loc[df["IndReca"] <= Q1, "IndReca"] = mean([Q1, Q2]) #Se haya y reemplaza los valores de IndReca que están por debajo del percentil 25% con el valor promedio entre 25-50%

In [ ]:
df.loc[df["IndReca"] >= Q3, "IndReca"] = mean([Q2, Q3]) #Se haya y reemplaza los valores de IndReca que están por encima del percentil 75% con el valor promedio entre 50-75%

Finalmente, valido que no tengo valores nulos con la variable valores_nulos y, que además, todos los atributos cumplan con las políticas de sustitución de inconsistencia usando la función describe()

In [ ]:
print(valores_nulos)

NIC         0
Estrato     0
Nhab        0
Ingresos    0
Consumo     0
IndReca     0
QSAT        0
dtype: int64


In [ ]:
df.describe()

NIC       Estrato          Nhab      Ingresos       Consumo  \
count  2.191700e+04  21917.000000  21917.000000  21917.000000  21917.000000   
mean   2.117604e+06      2.513711      4.006981      7.063782    304.436103   
std    5.108147e+05      1.111775      1.414729      4.089644     42.673752   
min    1.234686e+06      1.000000      2.000000      1.480000    235.000000   
25%    1.678706e+06      2.000000      3.000000      3.640000    272.000000   
50%    2.120192e+06      3.000000      4.000000      6.300000    302.000000   
75%    2.558549e+06      3.000000      5.000000     10.430000    344.000000   
max    2.999862e+06      4.000000      6.000000     16.730000    390.000000   

            IndReca          QSAT  
count  21917.000000  21917.000000  
mean       0.297894      2.949172  
std        0.046765      1.155652  
min        0.250000      1.000000  
25%        0.250000      2.000000  
50%        0.300000      2.900000  
75%        0.350000      3.900000  
max        0.380000      4.900000

Validando,

*   No tengo valores nulos.
*   Los predios de los Estratos tienen como valor máximo 4 y como valor mínimo 1, por lo que los valores se encuentran dentro del rango.
*   El Consumo tiene de valor máximo 390 kWh y como valor mínimo 235 kWh, y ya que estos rangos están entre 200 - 420 kWh, se cumplen con las políticas de sustitución de inconsistencias.
*   También, el QSAT tiene de valor máximo 4.9 y de valor mínimo 1.0, y ya que estos rangos están entre 1 - 5, se cumplen con las políticas de sustitución de inconsistencias.
*   Finalmente, los valores de IndReca están entre entre el promedio de 25-50% que es 0.2 y el 50-75% que es 0.4



In [ ]:
df.to_csv("ds-energy-consumers-depurado-AngelicaMoreno.csv", sep = ",", index = False)